File: android/src/main/java/io/invertase/firebase/links/RNFirebaseLinks.java
package io.invertase.firebase.links;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.util.Log;

import com.facebook.react.bridge.ActivityEventListener;
import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.appinvite.FirebaseAppInvite;
import com.google.firebase.dynamiclinks.DynamicLink;
import com.google.firebase.dynamiclinks.FirebaseDynamicLinks;
import com.google.firebase.dynamiclinks.PendingDynamicLinkData;
import com.google.firebase.dynamiclinks.ShortDynamicLink;

import javax.annotation.Nonnull;

import io.invertase.firebase.Utils;

public class RNFirebaseLinks extends ReactContextBaseJavaModule implements ActivityEventListener, LifecycleEventListener {
  private final static String TAG = RNFirebaseLinks.class.getCanonicalName();
  private String mInitialLink = null;
  private boolean mInitialLinkInitialized = false;

  public RNFirebaseLinks(ReactApplicationContext reactContext) {
    super(reactContext);
    getReactApplicationContext().addActivityEventListener(this);
    getReactApplicationContext().addLifecycleEventListener(this);
  }

  @Override
  public String getName() {
    return "RNFirebaseLinks";
  }

  @ReactMethod
  public void createDynamicLink(final ReadableMap linkData, final Promise promise) {
    try {
      DynamicLink.Builder builder = getDynamicLinkBuilder(linkData);
      String link = builder
        .buildDynamicLink()
        .getUri()
        .toString();
      Log.d(TAG, "created dynamic link: " + link);
      promise.resolve(link);
    } catch (Exception ex) {
      Log.e(TAG, "create dynamic link failure " + ex.getMessage());
      promise.reject("links/failure", ex.getMessage(), ex);
    }
  }

  @ReactMethod
  public void createShortDynamicLink(
    final ReadableMap linkData,
    final String type,
    final Promise promise
  ) {
    try {
      DynamicLink.Builder builder = getDynamicLinkBuilder(linkData);
      Task<ShortDynamicLink> shortLinkTask;
      if ("SHORT".equals(type)) {
        shortLinkTask = builder.buildShortDynamicLink(ShortDynamicLink.Suffix.SHORT);
      } else if ("UNGUESSABLE".equals(type)) {
        shortLinkTask = builder.buildShortDynamicLink(ShortDynamicLink.Suffix.UNGUESSABLE);
      } else {
        shortLinkTask = builder.buildShortDynamicLink();
      }

      shortLinkTask.addOnCompleteListener(new OnCompleteListener<ShortDynamicLink>() {
        @Override
        public void onComplete(@Nonnull Task<ShortDynamicLink> task) {
          if (task.isSuccessful()) {
            String shortLink = task
              .getResult()
              .getShortLink()
              .toString();
            Log.d(TAG, "created short dynamic link: " + shortLink);
            promise.resolve(shortLink);
          } else {
            Log.e(
              TAG,
              "create short dynamic link failure " + task
                .getException()
                .getMessage()
            );
            promise.reject(
              "links/failure",
              task
                .getException()
                .getMessage(),
              task.getException()
            );
          }
        }
      });
    } catch (Exception ex) {
      Log.e(TAG, "create short dynamic link failure " + ex.getMessage());
      promise.reject("links/failure", ex.getMessage(), ex);
    }
  }

  @ReactMethod
  public void getInitialLink(final Promise promise) {
    if (mInitialLinkInitialized) {
      promise.resolve(mInitialLink);
    } else {
      if (getCurrentActivity() != null) {
        FirebaseDynamicLinks
          .getInstance()
          .getDynamicLink(getCurrentActivity().getIntent())
          .addOnSuccessListener(new OnSuccessListener<PendingDynamicLinkData>() {
            @Override
            public void onSuccess(PendingDynamicLinkData pendingDynamicLinkData) {
              if (pendingDynamicLinkData != null
                && !isInvitation(pendingDynamicLinkData)) {

                mInitialLink = pendingDynamicLinkData
                  .getLink()
                  .toString();
              }
              Log.d(TAG, "getInitialLink: link is: " + mInitialLink);
              mInitialLinkInitialized = true;
              promise.resolve(mInitialLink);
            }
          })
          .addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@Nonnull Exception e) {
              Log.e(TAG, "getInitialLink: failed to resolve link", e);
              promise.reject("link/initial-link-error", e.getMessage(), e);
            }
          });
      } else {
        Log.d(TAG, "getInitialLink: activity is null");
        promise.resolve(null);
      }
    }
  }

  //////////////////////////////////////////////////////////////////////
  // Start ActivityEventListener methods
  //////////////////////////////////////////////////////////////////////
  @Override
  public void onActivityResult(Activity activity, int requestCode, int resultCode, Intent data) {
    // Not required for this module
  }

  @Override
  public void onNewIntent(Intent intent) {
    FirebaseDynamicLinks
      .getInstance()
      .getDynamicLink(intent)
      .addOnSuccessListener(new OnSuccessListener<PendingDynamicLinkData>() {
        @Override
        public void onSuccess(PendingDynamicLinkData pendingDynamicLinkData) {
          if (pendingDynamicLinkData != null
            && !isInvitation(pendingDynamicLinkData)) {
            String link = pendingDynamicLinkData
              .getLink()
              .toString();
            Utils.sendEvent(
              getReactApplicationContext(),
              "links_link_received",
              link
            );
          }
        }
      });
  }
  //////////////////////////////////////////////////////////////////////
  // End ActivityEventListener methods
  //////////////////////////////////////////////////////////////////////

  //////////////////////////////////////////////////////////////////////
  // Start LifecycleEventListener methods
  //////////////////////////////////////////////////////////////////////
  @Override
  public void onHostResume() {
    // Not required for this module
  }

  @Override
  public void onHostPause() {
    // Not required for this module
  }

  @Override
  public void onHostDestroy() {
    mInitialLink = null;
    mInitialLinkInitialized = false;
  }
  //////////////////////////////////////////////////////////////////////
  // End LifecycleEventListener methods
  //////////////////////////////////////////////////////////////////////

  // Looks at the internals of the link data to detect whether it's an invitation or not
  private boolean isInvitation(PendingDynamicLinkData pendingDynamicLinkData) {
    FirebaseAppInvite invite = FirebaseAppInvite.getInvitation(pendingDynamicLinkData);
    if (invite != null && invite.getInvitationId() != null && !invite
      .getInvitationId()
      .isEmpty()) {
      return true;
    }
    return false;
  }

  private DynamicLink.Builder getDynamicLinkBuilder(final ReadableMap linkData) {
    DynamicLink.Builder builder = FirebaseDynamicLinks
      .getInstance()
      .createDynamicLink();
    try {
      builder.setLink(Uri.parse(linkData.getString("link")));
      builder.setDomainUriPrefix(linkData.getString("domainURIPrefix"));
      setAnalyticsParameters(linkData.getMap("analytics"), builder);
      setAndroidParameters(linkData.getMap("android"), builder);
      setIosParameters(linkData.getMap("ios"), builder);
      setITunesParameters(linkData.getMap("itunes"), builder);
      setNavigationParameters(linkData.getMap("navigation"), builder);
      setSocialParameters(linkData.getMap("social"), builder);
    } catch (Exception e) {
      Log.e(TAG, "error while building parameters " + e.getMessage());
      throw e;
    }
    return builder;
  }

  private void setAnalyticsParameters(
    final ReadableMap analyticsData,
    final DynamicLink.Builder builder
  ) {
    DynamicLink.GoogleAnalyticsParameters.Builder analyticsParameters = new DynamicLink.GoogleAnalyticsParameters.Builder();

    if (analyticsData.hasKey("campaign")) {
      analyticsParameters.setCampaign(analyticsData.getString("campaign"));
    }
    if (analyticsData.hasKey("content")) {
      analyticsParameters.setContent(analyticsData.getString("content"));
    }
    if (analyticsData.hasKey("medium")) {
      analyticsParameters.setMedium(analyticsData.getString("medium"));
    }
    if (analyticsData.hasKey("source")) {
      analyticsParameters.setSource(analyticsData.getString("source"));
    }
    if (analyticsData.hasKey("term")) {
      analyticsParameters.setTerm(analyticsData.getString("term"));
    }
    builder.setGoogleAnalyticsParameters(analyticsParameters.build());
  }

  private void setAndroidParameters(
    final ReadableMap androidData,
    final DynamicLink.Builder builder
  ) {
    if (androidData.hasKey("packageName")) {
      DynamicLink.AndroidParameters.Builder androidParameters = new DynamicLink.AndroidParameters.Builder(
        androidData.getString("packageName"));

      if (androidData.hasKey("fallbackUrl")) {
        androidParameters.setFallbackUrl(Uri.parse(androidData.getString("fallbackUrl")));
      }
      if (androidData.hasKey("minimumVersion")) {
        androidParameters.setMinimumVersion(Integer.parseInt(androidData.getString("minimumVersion")));
      }
      builder.setAndroidParameters(androidParameters.build());
    }
  }

  private void setIosParameters(final ReadableMap iosData, final DynamicLink.Builder builder) {
    if (iosData.hasKey("bundleId")) {
      DynamicLink.IosParameters.Builder iosParameters =
        new DynamicLink.IosParameters.Builder(iosData.getString("bundleId"));

      if (iosData.hasKey("appStoreId")) {
        iosParameters.setAppStoreId(iosData.getString("appStoreId"));
      }
      if (iosData.hasKey("customScheme")) {
        iosParameters.setCustomScheme(iosData.getString("customScheme"));
      }
      if (iosData.hasKey("fallbackUrl")) {
        iosParameters.setFallbackUrl(Uri.parse(iosData.getString("fallbackUrl")));
      }
      if (iosData.hasKey("iPadBundleId")) {
        iosParameters.setIpadBundleId(iosData.getString("iPadBundleId"));
      }
      if (iosData.hasKey("iPadFallbackUrl")) {
        iosParameters.setIpadFallbackUrl(Uri.parse(iosData.getString("iPadFallbackUrl")));
      }
      if (iosData.hasKey("minimumVersion")) {
        iosParameters.setMinimumVersion(iosData.getString("minimumVersion"));
      }
      builder.setIosParameters(iosParameters.build());
    }
  }

  private void setITunesParameters(
    final ReadableMap itunesData,
    final DynamicLink.Builder builder
  ) {
    DynamicLink.ItunesConnectAnalyticsParameters.Builder itunesParameters = new DynamicLink.ItunesConnectAnalyticsParameters.Builder();

    if (itunesData.hasKey("affiliateToken")) {
      itunesParameters.setAffiliateToken(itunesData.getString("affiliateToken"));
    }
    if (itunesData.hasKey("campaignToken")) {
      itunesParameters.setCampaignToken(itunesData.getString("campaignToken"));
    }
    if (itunesData.hasKey("providerToken")) {
      itunesParameters.setProviderToken(itunesData.getString("providerToken"));
    }
    builder.setItunesConnectAnalyticsParameters(itunesParameters.build());
  }

  private void setNavigationParameters(
    final ReadableMap navigationData,
    final DynamicLink.Builder builder
  ) {
    DynamicLink.NavigationInfoParameters.Builder navigationParameters = new DynamicLink.NavigationInfoParameters.Builder();

    if (navigationData.hasKey("forcedRedirectEnabled")) {
      navigationParameters.setForcedRedirectEnabled(navigationData.getBoolean(
        "forcedRedirectEnabled"));
    }
    builder.setNavigationInfoParameters(navigationParameters.build());
  }

  private void setSocialParameters(
    final ReadableMap socialData,
    final DynamicLink.Builder builder
  ) {
    DynamicLink.SocialMetaTagParameters.Builder socialParameters = new DynamicLink.SocialMetaTagParameters.Builder();

    if (socialData.hasKey("descriptionText")) {
      socialParameters.setDescription(socialData.getString("descriptionText"));
    }
    if (socialData.hasKey("imageUrl")) {
      socialParameters.setImageUrl(Uri.parse(socialData.getString("imageUrl")));
    }
    if (socialData.hasKey("title")) {
      socialParameters.setTitle(socialData.getString("title"));
    }
    builder.setSocialMetaTagParameters(socialParameters.build());
  }
}

File: ios/RNFirebase/links/RNFirebaseLinks.m
#import "RNFirebaseLinks.h"

#if __has_include(<FirebaseDynamicLinks/FirebaseDynamicLinks.h>)
#import <Firebase.h>
#import "RNFirebaseEvents.h"
#import "RNFirebaseUtil.h"

@implementation RNFirebaseLinks

static RNFirebaseLinks *theRNFirebaseLinks = nil;
static NSString *initialLink = nil;
static bool jsReady = FALSE;

+ (nonnull instancetype)instance {
    // If an event comes in before the bridge has initialised the native module
    // then we create a temporary instance which handles events until the bridge
    // and JS side are ready
    if (theRNFirebaseLinks == nil) {
        theRNFirebaseLinks = [[RNFirebaseLinks alloc] init];
    }
    return theRNFirebaseLinks;
}

RCT_EXPORT_MODULE();

- (id)init {
    self = [super init];
    if (self != nil) {
        NSLog(@"Setting up RNFirebaseLinks instance");
        // Set static instance for use from AppDelegate
        theRNFirebaseLinks = self;
    }
    return self;
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

// *******************************************************
// ** Start AppDelegate methods
// *******************************************************

- (BOOL)application:(UIApplication *)app
            openURL:(NSURL *)url
            options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options {
    FIRDynamicLink *dynamicLink = [[FIRDynamicLinks dynamicLinks] dynamicLinkFromCustomSchemeURL:url];
    if (dynamicLink && dynamicLink.url) {
        NSURL* url = dynamicLink.url;
        [self sendJSEvent:self name:LINKS_LINK_RECEIVED body:url.absoluteString];
        return YES;
    }
    return NO;
}

- (BOOL)application:(UIApplication *)application
continueUserActivity:(NSUserActivity *)userActivity
 restorationHandler:(void (^)(NSArray *))restorationHandler {
    if ([userActivity.activityType isEqualToString:NSUserActivityTypeBrowsingWeb]) {
        return [[FIRDynamicLinks dynamicLinks]
                handleUniversalLink:userActivity.webpageURL
                completion:^(FIRDynamicLink * _Nullable dynamicLink, NSError * _Nullable error) {
                    if (dynamicLink && dynamicLink.url && error == nil) {
                        NSURL* url = dynamicLink.url;
                        [self sendJSEvent:self name:LINKS_LINK_RECEIVED body:url.absoluteString];
                    } else {
                        NSLog(@"Failed to handle universal link: %@", userActivity.webpageURL);
                    }
                }];
    }
    return NO;
}
// *******************************************************
// ** Finish AppDelegate methods
// *******************************************************

- (void)sendLink:(NSString *)link {
    [self sendJSEvent:self name:LINKS_LINK_RECEIVED body:link];
}

// ** Start React Module methods **
RCT_EXPORT_METHOD(createDynamicLink:(NSDictionary *)linkData
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    @try {
        FIRDynamicLinkComponents *dynamicLink = [self buildDynamicLink:linkData];
        
        if (dynamicLink == nil) {
            reject(@"links/failure", @"Failed to create Dynamic Link", nil);
        } else {
            NSString *longLink = dynamicLink.url.absoluteString;
            NSLog(@"created long dynamic link: %@", longLink);
            resolve(longLink);
        }
    }
    @catch(NSException * e) {
        NSLog(@"create dynamic link failure %@", e);
        reject(@"links/failure",[e reason], nil);
    }
}

RCT_EXPORT_METHOD(createShortDynamicLink:(NSDictionary *)linkData
                  type:(NSString *)type
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    @try {
        FIRDynamicLinkComponents *components = [self buildDynamicLink:linkData];
        if (type) {
            FIRDynamicLinkComponentsOptions *options = [FIRDynamicLinkComponentsOptions options];
            if ([type isEqual: @"SHORT"]) {
                options.pathLength = FIRShortDynamicLinkPathLengthShort;
            } else if ([type isEqual: @"UNGUESSABLE"]) {
                options.pathLength = FIRShortDynamicLinkPathLengthUnguessable;
            }
            components.options = options;
        }
        [components shortenWithCompletion:^(NSURL *_Nullable shortURL, NSArray *_Nullable warnings, NSError *_Nullable error) {
            if (error) {
                NSLog(@"create short dynamic link failure %@", [error localizedDescription]);
                reject(@"links/failure", @"Failed to create Short Dynamic Link", error);
            } else {
                NSString *shortLink = shortURL.absoluteString;
                NSLog(@"created short dynamic link: %@", shortLink);
                resolve(shortLink);
            }
        }];
    }
    @catch(NSException * e) {
        NSLog(@"create short dynamic link failure %@", e);
        reject(@"links/failure",[e reason], nil);
    }
}

RCT_EXPORT_METHOD(getInitialLink:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    if (self.bridge.launchOptions[UIApplicationLaunchOptionsURLKey]) {
        NSURL* url = (NSURL*)self.bridge.launchOptions[UIApplicationLaunchOptionsURLKey];
        FIRDynamicLink *dynamicLink = [[FIRDynamicLinks dynamicLinks] dynamicLinkFromCustomSchemeURL:url];
        resolve(dynamicLink ? dynamicLink.url.absoluteString : initialLink);
    } else if (self.bridge.launchOptions[UIApplicationLaunchOptionsUserActivityDictionaryKey]
               && [self.bridge.launchOptions[UIApplicationLaunchOptionsUserActivityDictionaryKey][UIApplicationLaunchOptionsUserActivityTypeKey] isEqualToString:NSUserActivityTypeBrowsingWeb]) {
        NSDictionary *dictionary = self.bridge.launchOptions[UIApplicationLaunchOptionsUserActivityDictionaryKey];
        NSUserActivity* userActivity = (NSUserActivity*) dictionary[@"UIApplicationLaunchOptionsUserActivityKey"];
        [[FIRDynamicLinks dynamicLinks] handleUniversalLink:userActivity.webpageURL
                                                 completion:^(FIRDynamicLink * _Nullable dynamicLink, NSError * _Nullable error) {
                                                     if (error != nil){
                                                         NSLog(@"Failed to handle universal link: %@", [error localizedDescription]);
                                                         reject(@"links/failure", @"Failed to handle universal link", error);
                                                     } else {
                                                         NSString* urlString = dynamicLink ? dynamicLink.url.absoluteString : userActivity.webpageURL.absoluteString;
                                                         NSLog(@"initial link is: %@", urlString);
                                                         resolve(urlString);
                                                     }
                                                 }];
    } else {
        resolve(initialLink);
    }
}

RCT_EXPORT_METHOD(jsInitialised:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    jsReady = TRUE;
    resolve(nil);
}

// ** Start internals **

// Because of the time delay between the app starting and the bridge being initialised
// we catch any events that are received before the JS is ready to receive them
- (void)sendJSEvent:(RCTEventEmitter *)emitter name:(NSString *)name body:(id)body {
    if (emitter.bridge && jsReady) {
        [RNFirebaseUtil sendJSEvent:emitter name:name body:body];
    } else if (!initialLink) {
        initialLink = body;
    } else {
        NSLog(@"Multiple link events received before the JS links module has been initialised");
    }
}

- (FIRDynamicLinkComponents *)buildDynamicLink:(NSDictionary *)linkData {
    @try {
        NSURL *link = [NSURL URLWithString:linkData[@"link"]];
        FIRDynamicLinkComponents *components = [FIRDynamicLinkComponents componentsWithLink:link domain:linkData[@"dynamicLinkDomain"]];
        
        [self setAnalyticsParameters:linkData[@"analytics"] components:components];
        [self setAndroidParameters:linkData[@"android"] components:components];
        [self setIosParameters:linkData[@"ios"] components:components];
        [self setITunesParameters:linkData[@"itunes"] components:components];
        [self setNavigationParameters:linkData[@"navigation"] components:components];
        [self setSocialParameters:linkData[@"social"] components:components];
        
        return components;
    }
    @catch(NSException * e) {
        NSLog(@"error while building componets from meta data %@", e);
        @throw;
    }
}

- (void)setAnalyticsParameters:(NSDictionary *)analyticsData
                    components:(FIRDynamicLinkComponents *)components {
    FIRDynamicLinkGoogleAnalyticsParameters *analyticsParams = [FIRDynamicLinkGoogleAnalyticsParameters parameters];
    
    if (analyticsData[@"campaign"]) {
        analyticsParams.campaign = analyticsData[@"campaign"];
    }
    if (analyticsData[@"content"]) {
        analyticsParams.content = analyticsData[@"content"];
    }
    if (analyticsData[@"medium"]) {
        analyticsParams.medium = analyticsData[@"medium"];
    }
    if (analyticsData[@"source"]) {
        analyticsParams.source = analyticsData[@"source"];
    }
    if (analyticsData[@"term"]) {
        analyticsParams.term = analyticsData[@"term"];
    }
    components.analyticsParameters = analyticsParams;
}

- (void)setAndroidParameters:(NSDictionary *)androidData
                  components:(FIRDynamicLinkComponents *)components {
    if (androidData[@"packageName"]) {
        FIRDynamicLinkAndroidParameters *androidParams = [FIRDynamicLinkAndroidParameters parametersWithPackageName: androidData[@"packageName"]];
        
        if (androidData[@"fallbackUrl"]) {
            androidParams.fallbackURL = [NSURL URLWithString:androidData[@"fallbackUrl"]];
        }
        if (androidData[@"minimumVersion"]) {
            androidParams.minimumVersion = [androidData[@"minimumVersion"] integerValue];
        }
        components.androidParameters = androidParams;
    }
}

- (void)setIosParameters:(NSDictionary *)iosData
              components:(FIRDynamicLinkComponents *)components {
    if (iosData[@"bundleId"]) {
        FIRDynamicLinkIOSParameters *iOSParams = [FIRDynamicLinkIOSParameters parametersWithBundleID:iosData[@"bundleId"]];
        if (iosData[@"appStoreId"]) {
            iOSParams.appStoreID = iosData[@"appStoreId"];
        }
        if (iosData[@"customScheme"]) {
            iOSParams.customScheme = iosData[@"customScheme"];
        }
        if (iosData[@"fallbackUrl"]) {
            iOSParams.fallbackURL = [NSURL URLWithString:iosData[@"fallbackUrl"]];
        }
        if (iosData[@"iPadBundleId"]) {
            iOSParams.iPadBundleID = iosData[@"iPadBundleId"];
        }
        if (iosData[@"iPadFallbackUrl"]) {
            iOSParams.iPadFallbackURL = [NSURL URLWithString:iosData[@"iPadFallbackUrl"]];
        }
        if (iosData[@"minimumVersion"]) {
            iOSParams.minimumAppVersion = iosData[@"minimumVersion"];
        }
        components.iOSParameters = iOSParams;
    }
}

- (void)setITunesParameters:(NSDictionary *)itunesData
                 components:(FIRDynamicLinkComponents *)components {
    FIRDynamicLinkItunesConnectAnalyticsParameters *itunesParams = [FIRDynamicLinkItunesConnectAnalyticsParameters parameters];
    if (itunesData[@"affiliateToken"]) {
        itunesParams.affiliateToken = itunesData[@"affiliateToken"];
    }
    if (itunesData[@"campaignToken"]) {
        itunesParams.campaignToken = itunesData[@"campaignToken"];
    }
    if (itunesData[@"providerToken"]) {
        itunesParams.providerToken = itunesData[@"providerToken"];
    }
    components.iTunesConnectParameters = itunesParams;
}

- (void)setNavigationParameters:(NSDictionary *)navigationData
                     components:(FIRDynamicLinkComponents *)components {
    FIRDynamicLinkNavigationInfoParameters *navigationParams = [FIRDynamicLinkNavigationInfoParameters parameters];
    if (navigationData[@"forcedRedirectEnabled"]) {
        navigationParams.forcedRedirectEnabled = navigationData[@"forcedRedirectEnabled"];
    }
    components.navigationInfoParameters = navigationParams;
}

- (void)setSocialParameters:(NSDictionary *)socialData
                 components:(FIRDynamicLinkComponents *)components {
    FIRDynamicLinkSocialMetaTagParameters *socialParams = [FIRDynamicLinkSocialMetaTagParameters parameters];
    if (socialData[@"descriptionText"]) {
        socialParams.descriptionText = socialData[@"descriptionText"];
    }
    if (socialData[@"imageUrl"]) {
        socialParams.imageURL = [NSURL URLWithString:socialData[@"imageUrl"]];
    }
    if (socialData[@"title"]) {
        socialParams.title = socialData[@"title"];
    }
    components.socialMetaTagParameters = socialParams;
}

- (NSArray<NSString *> *)supportedEvents {
    return @[LINKS_LINK_RECEIVED];
}

+ (BOOL)requiresMainQueueSetup {
    return YES;
}

@end

#else
@implementation RNFirebaseLinks
@end
#endif
